internal/ad/ad.go

```diff
@@ -491,116 +491,155 @@ func (ad *AD) parseGPOs(ctx context.Context, gpos []gpo, objectClass ObjectClass
 			Rules: make(map[string][]entry.Entry),
 		}
 		r = append(r, gpoWithRules)
-		if err := func() error {
-			ad.downloadables[name].mu.RLock()
-			defer ad.downloadables[name].mu.RUnlock()
-			_ = ad.downloadables[name].testConcurrent
-
-			log.Debugf(ctx, "Parsing GPO %q", name)
-
-			// We need to consider the uppercase version of the name as well,
-			// which could occur in some of the default GPOs such as Default
-			// Domain Policy.
-			classes := []string{"User", "USER"}
-			if objectClass == ComputerObject {
-				classes = []string{"Machine", "MACHINE"}
+		if err = ad.parseGPO(ctx, name, url, keyFilterPrefix, objectClass, gpoWithRules); err != nil {
+			return r, err
+		}
+	}
+
+	return r, nil
+}
+
+func (ad *AD) parseGPO(ctx context.Context, name, url, keyFilterPrefix string, objectClass ObjectClass, gpoWithRules policies.GPO) error {
+	ad.downloadables[name].mu.RLock()
+	defer ad.downloadables[name].mu.RUnlock()
+	_ = ad.downloadables[name].testConcurrent
+
+	log.Debugf(ctx, "Parsing GPO %q of class %q", name, objectClass)
+
+	// We need to consider the uppercase version of the name as well,
+	// which could occur in some of the default GPOs such as Default
+	// Domain Policy.
+	classes := []string{"User", "USER"}
+	if objectClass == ComputerObject {
+		classes = []string{"Machine", "MACHINE"}
+	}
+
+	var err error
+	var f *os.File
+	for _, class := range classes {
+		var e error
+		var files []os.DirEntry
+
+		policyDir := filepath.Join(ad.sysvolCacheDir, "Policies", filepath.Base(url), class)
+		files, e = os.ReadDir(policyDir)
+
+		if e != nil {
+			if errors.Is(e, fs.ErrNotExist) {
+				log.Debugf(ctx, "Not found policy directory %q", policyDir)
+			} else {
+				log.Warningf(ctx, "Failed to read policy directory %q: %s", policyDir, e)
 			}
+		} else {
+			log.Debugf(ctx, "Success reading policy directory %q: found %d files", policyDir, len(files))
+		}
+
+		foundPolicy := false
```

Comment:
Maybe put that before the loop?

We are trying to align left, meaning that the happy path (the directory path is found, we list the dir, a filename matching it is found and we can parse it) is on the left, then any error is either a "let’s skip to next directory/potential match" or "an early return".

We still need we ensure we need to find one element then, so foundPolicy can be before the for loop if you see what I mean?

Reply:
Yes, It was a bug, changed it to be before the loop.